### ISSUE TITLE
Deref coercions do not work with blocks

### ISSUE THREAD

**Issue by barosl** [rust-lang/rust#26978]
It seems that the compiler handles a block differently when coercing a value.

``` rust
fn f(_: &str) {}

fn main() {
    let x = "Akemi Homura".to_owned();

    f(&x); // OK
    f(&(x)); // OK
    f(&{x}); // Error
}
```

[RFC 401](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md) says that a block with type `U` is also a target for coercion, so I think this behavior is a bug.

> blocks, if a block has type `U`, then the last expression in the block (if it
> is not semicolon-terminated) is a coercion site to `U`. This includes blocks
> which are part of control flow statements, such as `if`/`else`, if the block
> has a known type.

Also, the compiler seems to be able to coerce blocks using some "trivial" rules (e.g. `&mut T` -> `&T`).

``` rust
fn f(_: &i32) {}

fn main() {
    let x = &mut 42;

    f(x); // OK
    f((x)); // OK
    f({x}); // OK
}
```

So I guess this is more likely a problem of auto-deref.

**Comment by Gankro** [comment]
CC @nrc

**Comment by bluss** [comment]
As some users have discovered, `{x}.item` can be used to force a `x: &mut X` to move instead of reborrow. Is that related? We might not be allowed to break that.

**Comment by nrc** [comment]
I suspect this is something specific to deref coercions. I think it is 'just a bug', but I don't know how easy it is to fix.

**Comment by tbu-** [comment]
This blocks usage of `try!` together with deref coercions if I understand this bug correctly:

`let _: &Path = &try!(env::current_dir());` fails to compile.

**Comment by barosl** [comment]
@tbu- Yup, actually this issue was found by a Stack Overflow question [that tried the same thing](http://stackoverflow.com/questions/31361334/why-cant-i-take-the-reference-of-the-result-of-try-with).

**Comment by llogiq** [comment]
There is also [this rust-users thread](https://users.rust-lang.org/t/deref-coercion-not-working-for-if-expressions/2917/2) with a very similar problem.

**Comment by nrc** [comment]
triage: I-nominated

**Comment by nrc** [comment]
triaged because it is such an annoying bug. Recommend p-medium.

**Comment by nikomatsakis** [comment]
cc @eddyb

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by eddyb** [comment]
It would be very helpful to point out the error, which is:
```rust
<anon>:8:9: 8:10 error: mismatched types:
 expected `str`,
    found `collections::string::String`
(expected str,
    found struct `collections::string::String`) [E0308]
<anon>:8     f(&{x}); // Error
                 ^
```
So the problem could be that `str` propagated through the block to `x`, despite being unsized.
As blocks are always rvalues, maybe the "expected type" propagation is not handling the block case?

**Comment by eddyb** [comment]
Update: That was dumb of me: I did not realize this only ever happens with `str`.
Of course it does, [because I never implemented it](https://github.com/rust-lang/rust/blame/083dc5b12d6d29e5d88f6446664968d1f868562d/src/librustc_typeck/check/mod.rs#L3831).

**Comment by ldm0** [comment]
This PR should be reopened:
```rust
fn f(_: &i32) {}

fn main() {
    let x = Box::new(1i32);

    f(&x); // OK
    f(&(x)); // OK
    f(&{x}); // Error
}
```

**Comment by eddyb** [comment]
I should've renamed the title of this issue, I believe (based on my comments in this thread) that this issue was specifically about `str` not working the same was `[T]`.

I doubt your example ever worked, because the block gets `i32` as the expected type and *that* is `Sized`, so there's no way to rule it out as the type of the block.
